README.md
# gcloud-partner-api-docs

https://cpapi.gabia.com/docs

## Infomation

### nightwatch
http://nightwatchjs.org/

https://github.com/dwyl/learn-nightwatch
 
### Dependencies
- https://www.npmjs.com/package/nightwatch
- https://www.npmjs.com/package/chromedriver
- https://www.npmjs.com/package/selenium-server

### Apache Reverse Proxy

#### Edit Apache Config
```
$ vi /usr/local/apache/conf/extra/httpd-vhosts.conf

$ vi /usr/local/apache/conf/extra/httpd-ssl.conf
```

```
ProxyPass "/docs" "http://cpapi.gabia.com:9000"
ProxyPassReverse "/docs" "http://cpapi.gabia.com:9000"

ProxyPass "/v1" "https://cpapi.gabia.com:7000/v1"
ProxyPassReverse "/v1" "https://cpapi.gabia.com:7000/v1"
```

#### Check & Restart Apache
```
$ /usr/local/apache/bin/apachectl configtest

## Syntax OK
```

```
$ /usr/local/apache/bin/apachectl stop

$ /usr/local/apache/bin/apachectl start
```

```
$ ps -aef|grep http
```

## Usage
> Docs on https://cpapi.gabia.com/docs

```
$ cd /home/node/cpapi/docs
$ git clone https://gitlab.gabia.com/cloud/service/gcloud-partner-api-docs.git ./
$ npm run build
```

### Dev
```
$ npm run build:dev
```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.8.2] - 2023-01-06
### Changed
- 정산금 조회 API 에 `users` 프로퍼티 추가
- 실시간 요금 현황 API 에 `users` 프로퍼티 추가

## [1.8.1] - 2022-07-14
### Changed
- Postman Collection 1.8.1

## [1.8.0] - 2022-05-24
### Changed
- NAS 버전 정보 추가
## [1.7.1] - 2022-06-09
### Changed
- 일부 VM 사양 종료함에 따라 제공하는 사양 안내 추가

## [1.7.0] - 2022-04-04
### Added
- SSD NAS 스펙 추가

## [1.6.0] - 2020-08-25
### Changed
- 로드밸런서 xfp, proxyp 속성 추가

## [1.5.2] - 2020-07-30
### Changed
- 서버 생성 시 응답값에 비밀번호 속성 추가
- Callback URL 내용 보완

## [1.5.1] - 2020-05-20
### Changed
- 클라우드 파트너 도입 가이드 v1.1

## [1.5.0] - 2020-02-18
### Added
- g클라우드 서버(베이직/프리미엄 SSD) 추가
- SSD 타입의 g클라우드 스토리지 추가
- MR 템플릿 추가
### Changed
- bills API response 수정
    - 정산금 상세 조회 시 네트워크 사용량이 표기되지 않는 문제 수정
        - 기존의 키 (in_traffic_domestic, out_traffic_domestic, in_traffic_foreign, out_traffic_foreign, in_traffic, out_traffic)은 표시할 방법이 없으므로 deprecated. 항상 0으로 표시됨
        - use_traffic(과금 대상 트래픽), use_traffic_domestic(과금 대상 국내 트래픽), use_traffic_foreign(과금 대상 해외 트래픽) 으로 표기
    - 정산금 상세 조회 및 실시간 사용 요금 조회 시 image_storages 에는 image size 를 표기하도록 수정
    - 실시간 요금 조회 시 servers 및 images 에 서버 OS 종류 (Linux, Windows) 를 표시하도록 수정
    - 기타 오기 수정
- 서버 타입 표기 변경
    - SSD -> 로컬 SSD
    - 베이직/프리미엄 HDD -> 서버 타입이 베이직/프리미엄이고 스토리지 타입이 HDD인 서버
    - 베이직/프리미엄 SSD -> 서버 타입이 베이직/프리미엄이고 스토리지 타입이 SSD인 서버
### Deprecated
- `v1/lbs` API 지원 중단


## [1.4.5] - 2019-12-05
- Error Code 정리

## [1.4.4] - 2019-10-25
- 2nd 존 오픈

## [1.4.3] - 2019-10-16
- MONITORING API 문구 수정
    - 데이터 간격에 대한 설명 추가
    - 조회 기간에 따라 데이터 간격이 달라질 수 있음을 명시
    - 기타 오기 수정

## [1.4.2] - 2019-09-26
### Changed
- redoc@1.22.3
    - issue: https://gitlab.gabia.com/cloud/service/gcloud-partner-api-docs/issues/9

## [1.4.1] - 2019-09-19
### Changed
- user_id validation 수정
    - `^[\w\-]{3,16}$` -> `^[\w]{3,16}$`
    - 관련 문구 수정

## [1.4.0] - 2019-07-05
### Added
- [Certificates](https://cpapi.gabia.com/docs/#tag/Certificates)
    - SSL 인증서 CRUD API 추가
### Changed
- [Loadbalancers](https://cpapi.gabia.com/docs/#tag/LoadBalancers)
    - added `POST /loadbalancers`
    - added `GET /loadbalancers`
    - added `GET /loadbalancers/:lb_id`
    - added `PUT /loadbalancers/:lb_id`
    - added `DELETE /loadbalancers/:lb_id`
    - added `PUT /loadbalancers/:lb_id/ips/:ip_id`
    - added `POST /loadbalancers/:lb_id/firewalls/:fg_id`
    - added `DELETE /loadbalancers/:lb_id/firewalls/:fg_id`
    - added `PUT /loadbalancers/:lb_id/network`
    - added `GET /loadbalancers/:lb_id/monitoring/traffic`
    - added `GET /loadbalancers/:lb_id/monitoring/session`
    - deprecated `POST /lbs`
    - deprecated `GET /lbs`
    - deprecated `GET /lbs/:lb_id`
    - deprecated `PUT /lbs/:lb_id`
    - deprecated `DELETE /lbs/:lb_id`
    - deprecated `PUT /lbs/:lb_id/ips/:ip_id`
    - deprecated `PUT /lbs/:lb_id/setup`
    - deprecated `PUT /lbs/:lb_id/network`
    - deprecated `POST /lbs/:lb_id/ports`
    - deprecated `PUT /lbs/:lb_id/ports`
    - deprecated `DELETE /lbs/:lb_id/ports`
    - deprecated `POST /lbs/:lb_id/ports/:port/servers`
    - deprecated `PUT /lbs/:lb_id/ports/:port/servers`
    - deprecated `DELETE /lbs/:lb_id/ports/:port/servers`
    - deprecated `GET /lbs/:lb_id/monitoring/traffic`
    - deprecated `GET /lbs/:lb_id/monitoring/session`
- [Users](https://cpapi.gabia.com/docs/#tag/Users)
    - Responses
        - added `certificates`
- [Firewalls](https://cpapi.gabia.com/docs/#tag/Firewalls)
    - added `POST /loadbalancers/:lb_id/firewalls/:fg_id`
    - added `DELETE /loadbalancers/:lb_id/firewalls/:fg_id`
    - Responses
        - added `attached_lbs`
- [Monitoring](https://cpapi.gabia.com/docs/#tag/Monitoring)
    - added `GET /loadbalancers/:lb_id/monitoring/traffic`
    - added `GET /loadbalancers/:lb_id/monitoring/session`
    - deprecated `GET /lbs/:lb_id/monitoring/traffic`
    - deprecated `GET /lbs/:lb_id/monitoring/session`
- [Bills](https://cpapi.gabia.com/docs/#tag/Bills)
    - 로드밸런서 국내, 해외 트래픽 요금을 별도로 과금

## [1.3.2] - 2019-06-05
### Changed
- API 가이드 - 비동기작업 내용 수정

## [1.3.1] - 2019-05-14
### Changed
- Snapshots, Images, Server(삭제) 관련 문구 수정

## [1.3.0] - 2019-03-04
### Changed
- [Users](https://cpapi.gabia.com/docs/#tag/Users)
    - `cloud`라는 사용자 생성 불가
    - 파트너 ID와 동일한 사용자 생성 불가

- [Servers](https://cpapi.gabia.com/docs/#tag/Servers)
    - removed `POST /servers/:svr_id/fws`
    - removed `DELETE /servers/:svr_id/fws`
    - Responses
        - removed `fws`

- [Images](https://cpapi.gabia.com/docs/#tag/Images)
    - 파트너 ID와 동일한 사용자가 만든 이미지 자동으로 파트너의 모든 사용자에게 공유
        - 이미지 생성 후 추가되는 사용자도 자동으로 공유 가능

## [1.2.1] - 2019-01-18
### Changed
- [Servers](https://cpapi.gabia.com/docs/#tag/Servers)
    - Responses
        - added `firewalls` 
        - deprecated `fws` 

- [Users](https://cpapi.gabia.com/docs/#tag/Users)
    - Responses
        - added `nas` 
        - added `nas_snapshots` 
        - added `firewalls` 

- [Bills](https://cpapi.gabia.com/docs/#tag/Bills)
    - Path Parameters
        - added `bill_date`
        - added `user_id`
    - Responses
        - added `nas_total_charge` 
        - added `nas_snapshot_total_charge` 
        - added `nas` 
        - added `nas_snapshots` 

## [1.2.0] - 2019-01-17

`RELEASED` on [github](https://github.com/gcloudService/API-Partners-docs/releases/tag/1.2.0)

### Added
- [NAS](https://cpapi.gabia.com/docs/#tag/NAS)
- [Firewalls](https://cpapi.gabia.com/docs/#tag/Firewalls)
### Changed
- [operationId](https://swagger.io/specification/#operationObject) 추가

#### API
- [createServer](https://cpapi.gabia.com/docs/#operation/createServer)
    - `fg_id` 파라미터 추가

#### deprecated

> `방화벽 정책 추가`, `방화벽 정책 삭제` API는 더 이상 유효하지 않습니다.
>
> 기존 API와의 호환을 위해 2019.03.03 까지 유지하고 해당 API는 이용중지 됩니다.
> 
> 유지 기간 동안 해당 API를 이용 시 내부적으로 `서버 방화벽 그룹 설정`, `서버 방화벽 그룹 해제`를 호출합니다.



test/server.js
function makeServer(done) {
	var express = require('express')
	var app = express()

	app.use(express.static('./'))
	
	var server = app.listen(9000, function () {
		var port = server.address().port
		done()
	})
	return server
}
module.exports = makeServer;

test/e2e/index.js
var config = require('../../nightwatch/nightwatch.conf.BASIC.js')

module.exports = { // adapted from: https://git.io/vodU0
	'Guinea Pig Assert Title': function(browser) {
		browser
			.url('https://saucelabs.com/test/guinea-pig')
			.waitForElementVisible('body')
			.assert.title('I am a page title - Sauce Labs')
			.saveScreenshot('guinea-pig-test.png')
			.end()
	}
}

test/cpapi.gabia.com/docs.js
var config = require('../../nightwatch/nightwatch.conf.js')
var pack = require('../../package.json')



module.exports = { // adapted from: https://git.io/vodU0
	before: function (browser, done) {
		server = require('../server')(done) // done is a callback that executes when the server is started
	},

	after: function () {
		server.close()
	},

	'/docs': function (browser) {
		browser
			.url('http://localhost:9000')   // visit the local url
			.windowSize('windowHandle', 1200, 960)
			.waitForElementNotVisible('loading-bar', 10 * 1000)
			.saveScreenshot(config.imgpath(browser) + 'index.png')

		browser
			.expect.element('.api-info-version').text.to.contain(pack.version)	
	
		browser
			.assert.title('클라우드 파트너 API') // assert contains
			.end()
		

	}
}

nightwatch/nightwatch.conf.BASIC.js
require('env2')('.env') // optionally store your Evironment Variables in .env
const seleniumServer = require('selenium-server')
const chromedriver = require('chromedriver')
const SCREENSHOT_PATH = '../screenshots/'

// we use a nightwatch.conf.js file so we can include comments and helper functions
module.exports = {
	'src_folders': [
		'test/e2e'// Where you are storing your Nightwatch e2e tests
	],
	'output_folder': './reports', // reports (test outcome) output by nightwatch
	'selenium': {
		'start_process': true, // tells nightwatch to start/stop the selenium process
		'server_path': seleniumServer.path,
		'host': '127.0.0.1',
		'port': 4444, // standard selenium port
		'cli_args': {
			'webdriver.chrome.driver' : chromedriver.path
		}
	},
	'test_settings': {
		'default': {
			'screenshots': {
				'enabled': true, // if you want to keep screenshots
				'path': SCREENSHOT_PATH // save screenshots here
			},
			'globals': {
				'waitForConditionTimeout': 5000 // sometimes internet is slow so wait.
			},
			'desiredCapabilities': { // use Chrome as the default browser for tests
				'browserName': 'chrome'
			}
		},
		'chrome': {
			'desiredCapabilities': {
				'browserName': 'chrome',
				'javascriptEnabled': true // turn off to test progressive enhancement
			}
		}
	}
}

function padLeft (count) { // theregister.co.uk/2016/03/23/npm_left_pad_chaos/
	return count < 10 ? '0' + count : count.toString()
}

var FILECOUNT = 0 // 'global' screenshot file count
/**
 * The default is to save screenshots to the root of your project even though
 * there is a screenshots path in the config object above! ... so we need a
 * function that returns the correct path for storing our screenshots.
 * While we're at it, we are adding some meta-data to the filename, specifically
 * the Platform/Browser where the test was run and the test (file) name.
 */
function imgpath (browser) {
	var a = browser.options.desiredCapabilities
	var meta = [a.platform]
	meta.push(a.browserName ? a.browserName : 'any')
	meta.push(a.version ? a.version : 'any')
	meta.push(a.name) // this is the test filename so always exists.
	var metadata = meta.join('~').toLowerCase().replace(/ /g, '')
	return SCREENSHOT_PATH + metadata + '_' + padLeft(FILECOUNT++) + '_'
}

module.exports.imgpath = imgpath
module.exports.SCREENSHOT_PATH = SCREENSHOT_PATH

nightwatch/nightwatch.conf.js
require('env2')('.env') // optionally store your Evironment Variables in .env
const seleniumServer = require('selenium-server')
const chromedriver = require('chromedriver')
const SCREENSHOT_PATH = '../screenshots/'

// we use a nightwatch.conf.js file so we can include comments and helper functions
module.exports = {
	'src_folders': [
		'test/cpapi.gabia.com'// Where you are storing your Nightwatch e2e tests
	],
	'output_folder': '../reports', // reports (test outcome) output by nightwatch
	'selenium': {
		'start_process': true, // tells nightwatch to start/stop the selenium process
		'server_path': seleniumServer.path,
		'host': '127.0.0.1',
		'port': 4444, // standard selenium port
		'cli_args': {
			'webdriver.chrome.driver' : chromedriver.path
		}
	},
	'test_settings': {
		'default': {
			'screenshots': {
				'enabled': true, // if you want to keep screenshots
				'path': SCREENSHOT_PATH // save screenshots here
			},
			'globals': {
				'waitForConditionTimeout': 5000 // sometimes internet is slow so wait.
			},
			'desiredCapabilities': { // use Chrome as the default browser for tests
				'browserName': 'chrome'
			}
		},
		'chrome': {
			'desiredCapabilities': {
				'browserName': 'chrome',
				'javascriptEnabled': true // turn off to test progressive enhancement
			}
		}
	}
}

function padLeft (count) { // theregister.co.uk/2016/03/23/npm_left_pad_chaos/
	return count < 10 ? '0' + count : count.toString()
}

var FILECOUNT = 0 // 'global' screenshot file count
/**
 * The default is to save screenshots to the root of your project even though
 * there is a screenshots path in the config object above! ... so we need a
 * function that returns the correct path for storing our screenshots.
 * While we're at it, we are adding some meta-data to the filename, specifically
 * the Platform/Browser where the test was run and the test (file) name.
 */
function imgpath (browser) {
	var a = browser.options.desiredCapabilities
	var meta = [a.platform]
	meta.push(a.browserName ? a.browserName : 'any')
	meta.push(a.version ? a.version : 'any')
	meta.push(a.name) // this is the test filename so always exists.
	var metadata = meta.join('~').toLowerCase().replace(/ /g, '')
	return SCREENSHOT_PATH + metadata + '_' + padLeft(FILECOUNT++) + '_'
}

module.exports.imgpath = imgpath
module.exports.SCREENSHOT_PATH = SCREENSHOT_PATH